site/book/05-developing-functions/02-developing-in-Go.md

```diff
@@ -1,133 +1,177 @@
+  if err := fn.AsMain(fn.ResourceListProcessorFunc(Run)); err != nil {
+    os.Exit(1)
+  }
 }
 ```
 
-### Build and test the function
+The [`fn`] library provides a series of KRM level operations for [`ResourceList`]. 
+Basically, the KRM resource `ResourceList.FunctionConfig` and KRM resources `ResourceList.Items` are both converted to 
+`KubeObject` objects. You can use `KubeObject` similar as [`unstructured.Unstrucutred`].
 
-Build the go binary:
+The set-annotation function (see below) iterates the `ResourceList.Items`, finds out the `Deployment` resources and
+adds the annotation. After the iteration, it adds some user message to the `ResourceList.Results`
 
-```shell
-$ go mod tidy; go build -o my-fn .
+```go
+func Run(rl *fn.ResourceList) (bool, error) {
+    for _, kubeObject := range rl.Items {
+        if kubeObject.IsGVK("v1", "Deployment") {
+            kubeObject.SetAnnotation("config.kubernetes.io/managed-by", "kpt")
+        }
+    }
+    rl.Results = append(rl.Results, fn.GeneralResult("Add config.kubernetes.io/managed-by=kpt to all `Deployment` resources", fn.Info))
+    return true, nil
+}
 ```
 
-Fetch the wordpress package:
+Learn more about the `KubeObject` from the [go doc](https://pkg.go.dev/github.com/GoogleContainerTools/kpt-functions-sdk/go/fn)
 
-```shell
-$ kpt pkg get https://github.com/GoogleContainerTools/kpt.git/package-examples/wordpress@v0.8
-```
 
-Test it by running the function imperatively:
+### Test the KRM function
+
+The "get-started" package contains a `./data` directory. You can use this to test out your functions. 
 
 ```shell
-$ kpt fn eval wordpress --exec ./my-fn -- myannotation=foo
-```
+# Edit the `data/resources.yaml` with your KRM resources. 
+# resources.yaml already has a `Deployment` and `Service` as test data. 
+vim data/resources.yaml
 
-During iterative development, `--exec` flag can be used to execute the
-function binary directly instead of requiring the function to be containerized
-first. Once you have a function binary that works, you can then proceed to
-creating the container image.
+# Convert the KRM resources and FunctionConfig resource to `ResourceList`.
+kpt fn source data
+
+# Pass the ResourceList as StdIn to your function
+kpt fn source data | go run main.go
 
-### Publish the function
+# Verify the KRM function behavior in the StdOutput `ResourceList`
+```
 
-Generate a Dockerfile for the function image:
+### Build the KRM function as an executable
 
+After editing your function, you can release it as an executable
 ```shell
-$ go run ./main.go gen ./
+go build -o ${FUNCTION_NAME}
+# Change permission mode if needed
+chmod +x ${FUNCTION_NAME}
 ```
 
-Build the image:
-
+#### Run the executable imperatively via kpt
 ```shell
-$ docker build . -t gcr.io/project/fn-name:tag
+kpt fn eval ./data --exec ./${FUNCTION_NAME} 
 ```
 
-Optionally, push the image to a container registry:
+#### Run the executable declaratively via kpt
```

Comment:
hmm..we don't encourage users to run executables declaratively because it leads to non-portable packages, so not sure if we want to put this upfront here.

Reply:
I have to say it is really hard to guide users from the authors perspective.
The main reason I introduce the executable is that from my own experience, this is the missing piece that I cannot quickly understand the SDK (based on kyaml). When debugging the function, I need a quick and easy way to run the function, and the `docker build`+`kpt fn render` is a terrible experience, especially it ignore errors if not panic or append to results.

I'll call it out that the executable is an easy approach for developing. And the docker image is the right approach for fn usage (releasable) WDYT?

Reply:
You are right. I reran the user guide again and decided to remove the entire executable section. It feels to be duplicate experiences (the same function is run 3 times with different kpt commands)  and not the main focus of the guide. Thanks!